solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/health/indicators/FlowHealthIndicator.java

```diff
@@ -0,0 +1,52 @@
+package com.solace.spring.cloud.stream.binder.health.indicators;
+
+import com.solace.spring.cloud.stream.binder.health.base.SolaceHealthIndicator;
+import com.solace.spring.cloud.stream.binder.properties.SolaceFlowHealthProperties;
+import com.solacesystems.jcsmp.FlowEventArgs;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.lang.Nullable;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class FlowHealthIndicator extends SolaceHealthIndicator {
+	private final SolaceFlowHealthProperties solaceFlowHealthProperties;
+	private final AtomicLong reconnectCount = new AtomicLong(0);
+	private static final Log logger = LogFactory.getLog(FlowHealthIndicator.class);
+
+	public FlowHealthIndicator(SolaceFlowHealthProperties solaceFlowHealthProperties) {
+		this.solaceFlowHealthProperties = solaceFlowHealthProperties;
+	}
+
+	public void up() {
+		super.healthUp();
+	}
+
+	public void reconnecting(@Nullable FlowEventArgs eventArgs) {
+		long reconnectAttempt = this.reconnectCount.incrementAndGet();
+		if (Optional.of(this.solaceFlowHealthProperties.getReconnectAttemptsUntilDown())
+				.filter(maxReconnectAttempts -> maxReconnectAttempts > 0)
+				.filter(maxReconnectAttempts -> reconnectAttempt > maxReconnectAttempts)
+				.isPresent()) {
+			if (logger.isDebugEnabled()) {
+				logger.debug(String.format("Solace flow reconnect attempt %s > %s, changing state to down",
+						reconnectAttempt, this.solaceFlowHealthProperties.getReconnectAttemptsUntilDown()));
+			}
+			this.down(eventArgs);
+			this.reconnectCount.set(0);
```

Comment:
What happens if the flow goes into `RECONNECTING` -> `UP` -> `RECONNECTING`?

Looking at the implementation, it seems to me that the 2nd `RECONNECTING` event will start with `reconnectCount=1` instead of `reconnectCount=0`.

Reply:
oh yeah, thank you for that, that's definitely a case I missed. Any state change should reset the counter, so I added it to the `this.up` and `this.down` methods as well, as the sequence might be `reconnecting` -> `down` (not withing this method, but as a separate event) -> `whatever the next...`